QueueDown/MultiStagePipes.cs

```diff
@@ -0,0 +1,95 @@
+using System.Buffers;
+using System.Collections.Immutable;
+using System.IO.Pipelines;
+
+namespace QueueDown;
+
+public static class MultiStagePipes
+{
+    /// <summary>
+    /// Basically tries to reduce write contention but giving each thread/worker
+    /// it's own pipe to write to. Then a single reader pushes input into the
+    /// output pipe.
+    /// </summary>
+    public static void Run(Pipe outputPipe, MemoryPool<byte> pool)
+    {
+        PipeGroup readers = new();
+
+        for (int i = 0; i < 50; i++) {
+            var __ = Task.Run(async () => {
+                var producerPipe = new Pipe(new(pool));
+                readers.AddPipe(producerPipe.Reader);
+
+
+                // Write to the pipe
+                while (true) {
+                    var buffer = producerPipe.Writer.GetMemory();
+                    buffer.Span.Fill((byte)i);
+                    producerPipe.Writer.Advance(buffer.Length);
+                    await producerPipe.Writer.FlushAsync();
+                }
+            });
+        }
+
+        var _ = Task.Run(() => ReadAllPipes(readers, outputPipe.Writer));
+    }
+
+    /// <summary>
+    /// Reads from pipes in a round-robin manner and pushes output into final pipe
+    /// </summary>
+    private static async Task ReadAllPipes(PipeGroup stage1Pipes, PipeWriter outputPipe)
+    {
+        while (true) {
+            var pipes = stage1Pipes.Readers;
+
+            foreach (var inputPipe in pipes) 
```

Comment:
This would melt the CPU though right? What happens if there's no data in the input pipes? This just loops aggressively and doesn't wait.

Reply:
Like I said, this is rough 😁

I could try a WaitAny as a backoff if we complete a loop without any successful reads